src/result_processor.c

```diff
@@ -553,32 +553,41 @@ void SortAscMap_Dump(uint64_t tt, size_t n) {
 typedef struct {
   ResultProcessor base;
   uint32_t offset;
-  uint32_t limit;
-  uint32_t count;
+  uint32_t remaining;
 } RPPager;
 
-static int rppagerNext(ResultProcessor *base, SearchResult *r) {
+static int rppagerNext_Limit(ResultProcessor *base, SearchResult *r) {
   RPPager *self = (RPPager *)base;
-  int rc;
 
+  // If we've reached LIMIT:
+  if (!self->remaining) {
+    return RS_RESULT_EOF;
+  }
+
+  self->remaining--;
+  return base->upstream->Next(base->upstream, r);
+}
+
+static int rppagerNext_Skip(ResultProcessor *base, SearchResult *r) {
+  RPPager *self = (RPPager *)base;
+
+  // Currently a pager is never called more than offset+limit times.
+  // We limit the entire pipeline to offset+limit (upstream and downstream).
+  uint32_t limit = MIN(self->remaining, base->parent->resultLimit);
+  base->parent->resultLimit = self->offset + limit;
   // If we've not reached the offset
-  while (self->count < self->offset) {
+  while (self->offset) {
     int rc = base->upstream->Next(base->upstream, r);
     if (rc != RS_RESULT_OK) {
       return rc;
     }
-    self->count++;
+    base->parent->resultLimit--;
```

Comment:
We do not get here with LIMIT 0 0, right?
Can both `limit` and `self->offset` be `0` somehow?
(wrapping to max `uint32_t`)

Reply:
we don't get here on `LIMIT 0 0` as these parameters cause the creation of a counter.
But even if we did, we decrease `base->parent->resultLimit` and `self->offset` only if `self->offset` is not 0, and `base->parent->resultLimit` is at least (>=) `self->offset`